### PR TITLE
Add get_processed_inner_instruction syscall

### LINKED ISSUE #22437: Add ability for programs to check the processed instruction stack
#### Problem

There are certain scenarios where instructions (top-level or via CPI) would like to know what the last instruction processed was.  Some examples are verification of assert instructions or to check memo instructions

#### Proposed Solution

Make at least the last executed instruction available to programs.  With the option of expanding to to a list of all past processed instructions in the tx.  There has already been work done in this general direction so hopefully, this will not be very disruptive.

### PR DESCRIPTION
**Pull request by jackcmay**
#### Problem

There are scenarios where instructions (top-level or via CPI) would like to know what the last instruction processed was. Some examples are verification of assert instructions or to check for memo instructions.  Programs can already check tx level instructions via the `Sysvar1nstructions1111111111111111111111111` but there is no way to get information about what inner instructions have been processed.

#### Summary of Changes

Add a syscall that allows a program to query what inner instructions have been successfully processed.

Fixes #22437

### CODE REVIEW

File: programs/bpf/tests/programs.rs
Comment (CriesofCarrots) [inline programs/bpf/tests/programs.rs:234]: Did cargo fmt do this? Wonder why...

File: program-runtime/src/invoke_context.rs
Comment (CriesofCarrots) [inline program-runtime/src/invoke_context.rs:1020]: I might be missing something obvious, but I don't see where this is called...?

File: programs/bpf/rust/sibling_instruction/src/lib.rs
Comment (CriesofCarrots) [inline programs/bpf/rust/sibling_instruction/src/lib.rs:58]: This usage looks great for my use-case

File: programs/bpf_loader/src/syscalls.rs
Comment (jackcmay) [inline programs/bpf_loader/src/syscalls.rs:2662]: @CriesofCarrots Last min rename, called here, i'll fixup

File: programs/bpf/tests/programs.rs
Comment (jackcmay) [reply programs/bpf/tests/programs.rs:236]: Yeah, dunno why

**Review by CriesofCarrots — APPROVED** [review]
For my part, looks good once CI is happy

File: program-runtime/src/invoke_context.rs
Comment (t-nelson) [inline program-runtime/src/invoke_context.rs:1025]: This can only access _prior executed_ sibling instructions, right?  Perhaps we can make the function name communicate that?

Comment (jackcmay) [reply program-runtime/src/invoke_context.rs:1025]: Something like `get_processed_sibling_instructions`?  Any suggestions?

Comment (t-nelson) [reply program-runtime/src/invoke_context.rs:1025]: `processed` might get confused with commitment somehow :thinking: 

Maybe `prior`, `previous`, `preceding`?  @CriesofCarrots, halp!

Comment (jackcmay) [reply program-runtime/src/invoke_context.rs:1025]: `get_succeeded_sibling_instructions`?
`get_completed_sibling_instructions`?

Comment (jackcmay) [reply program-runtime/src/invoke_context.rs:1025]: any extra in the name should reinforce that the instruction was successful and completed

Comment (CriesofCarrots) [reply program-runtime/src/invoke_context.rs:1025]: I actually don't mind `processed`, what with `process_instruction`. And that is essentially what processed commitment is referring to. However, if you think that's ambiguous, I also like `completed`

Comment (CriesofCarrots) [reply program-runtime/src/invoke_context.rs:1025]: I suppose `executed` is another option (referring to Justin's refactor here: https://github.com/solana-labs/solana/pull/22215, which refers to transactions, but similar meaning)

Comment (jackcmay) [reply program-runtime/src/invoke_context.rs:1025]: Hmm, I'd like to keep the `processed` terminology when it comes to instructions, especially due to `process_instruction`.

File: programs/bpf_loader/src/syscalls.rs
Comment (Lichtso) [inline programs/bpf_loader/src/syscalls.rs:2662]: Two things:
- Can we use `CompiledInstruction` which has account indices, instead of `Pubkey`s? (better for ABIv2).
- How about referencing the recorded instructions by index / a range?

Comment (jackcmay) [reply programs/bpf_loader/src/syscalls.rs:2662]: Instructions aren't always recorded, are you thinking we should always turn on full instruction recording?

I'll take a look at using `CompiledInstruction`

Comment (Lichtso) [reply programs/bpf_loader/src/syscalls.rs:2662]: [This refactoring](https://github.com/solana-labs/solana/pull/22111) should have made it cheap enough to always run it. And this PR here comes close to "always on" as well. So why not have them both use an unified underlying mechanism?

Comment (jackcmay) [reply programs/bpf_loader/src/syscalls.rs]: Yeah, I'll take a look at that, this implementation is pretty cheap as is, but we can probably do better

Comment (jackcmay) [reply programs/bpf_loader/src/syscalls.rs]: One thing to keep in mind is that the instruction recorder records instructions before they are invoked, so it includes instructions that haven't been fully processed yet which also results in a different order

Comment (jackcmay) [reply programs/bpf_loader/src/syscalls.rs]: Recorded inner instructions are defined being invoked, not successfully processed:
https://docs.solana.com/developing/clients/jsonrpc-api#inner-instructions-structure

File: program-runtime/src/invoke_context.rs
Comment (CriesofCarrots) [reply program-runtime/src/invoke_context.rs]: Cool, as I said, I'm fine with `processed`

Comment (t-nelson) [reply program-runtime/src/invoke_context.rs]: I was hoping for something that more explicitly indicated the positional restriction, I suppose it's implied by processed.  Probably going to be a point of confusion until the community knowledge base gets the hang of it.

Comment (jackcmay) [reply program-runtime/src/invoke_context.rs]: Another rename, doesn't really address your concern but I think it makes the terminology more consistent with previous work.  Open to suggestions for how to convey the positional restrictions.

Comment (CriesofCarrots) [reply program-runtime/src/invoke_context.rs]: The rename suggests that I misunderstood the PR description and that this syscall will *not* work for tx-level instructions, only between CPI ix within the same invoking ix. Is that correct?

Comment (jackcmay) [reply program-runtime/src/invoke_context.rs]: That is correct, just inner instructions within the same top level instruction.

Comment (CriesofCarrots) [reply program-runtime/src/invoke_context.rs]: Ah, okay. Bit of a bummer to have to use different apis depending on call depth, but can make it work. Thx!

File: sdk/program/src/instruction.rs
Comment (mvines) [inline sdk/program/src/instruction.rs:668]: I think it would be quite helpful to return the depth of the processed inner instructions as well, perhaps by changing the return type to `<Option<(usize, Instruction)>>`.

For example, `E` may want to confirm that `C` immediately preceded it and not care about what instructions `C` internally invoked (which may be `D` or otherwise depending on the arguments into `C`).  `E` currently has no way to figure that `D` was not at its invoke depth.

Comment (CriesofCarrots) [reply sdk/program/src/instruction.rs:668]: +1

Comment (jackcmay) [reply sdk/program/src/instruction.rs:668]: Interesting, programs don't currently have any info on what depth they are being invoked.  We could either return a relative depth, or provide another syscall that provides the current depth.  The latter might be useful in other situations.

File: sdk/program/src/instruction.rs
Comment (jackcmay) [inline sdk/program/src/instruction.rs:669]: @CriesofCarrots Can you double-check that this meets your use case

File: programs/bpf/rust/inner_instruction/src/lib.rs
Comment (jackcmay) [inline programs/bpf/rust/inner_instruction/src/lib.rs:5]: Also, @CriesofCarrots here is a usage example/test

File: sdk/program/src/instruction.rs
Comment (mvines) [reply sdk/program/src/instruction.rs]: > programs don't currently have any info on what depth they are being invoked

Native programs do via `InvokeContext::invoke_depth()`, it's been useful from time to time so exposing that into BPF programs via syscall as well seems quite nice.

Comment (jackcmay) [reply sdk/program/src/instruction.rs]: https://github.com/solana-labs/solana/pull/22515/commits/02a7f3d145f345dcb6def51d9a1a9075e159214e

File: sdk/program/src/instruction.rs
Comment (CriesofCarrots) [reply sdk/program/src/instruction.rs:669]: Thanks, @jackcmay , this will work

File: sdk/program/src/instruction.rs
Comment (mvines) [inline sdk/program/src/instruction.rs:653]: nit: `InnerMeta` seems a little too generic.  `ProcessedInnerInstructionMeta`?

**Review by mvines — APPROVED** [review]

File: program-runtime/src/invoke_context.rs
Comment (Lichtso) [inline program-runtime/src/invoke_context.rs:200]: This should go in `TransactionContext`, because it is not runtime internal and should be exposed to ABIv2 programs as well. I am preparing a PR that moves the `InstructionRecorder` there, on which you could rebase this PR.

File: programs/bpf_loader/src/syscalls.rs
Comment (Lichtso) [inline programs/bpf_loader/src/syscalls.rs:3068]: Lets convert from `CompiledInstruction` to `Instruction` on demand here.

**Review by Lichtso — CHANGES_REQUESTED** [review]

File: program-runtime/src/invoke_context.rs
Comment (jackcmay) [reply program-runtime/src/invoke_context.rs:200]: I'm going to merge this today, feel free to move it after

File: programs/bpf_loader/src/syscalls.rs
Comment (jackcmay) [reply programs/bpf_loader/src/syscalls.rs:3068]: As we discussed, we can make those changes after this pr lands.

**Comment by jackcmay** [comment]
@mvines @CriesofCarrots, @Lichtso has suggested an ABI change, what do you think?

Instead of returning a reverse-ordered list of successfully processed instructions in the form of (depth, instruction):
```
/// A
/// B -> C -> D
/// B -> E
/// B -> F
///
/// Then B's processed inner instruction list is: [(1, F), (1, E), (1, C), (2, D)]
```
 we could return the list of inner instructions in the same way we report recorded instructions, which means a reverse-ordered list of instructions in the form of (depth, status, instruction)
```
/// A
/// B -> C -> D
/// B -> E
/// B -> F
///
/// Then B's processed inner instruction list is: [(1, Success, F), (1, Success, E), (2, Success, D), (1, Success, C), (0, InProgress, B)]
```

**Comment by mvines** [comment]
> (depth, status, instruction)

The status field seems unhelpful.  The only way that B could be `InProgress` at that time is due to the `Success`-ful result of the previous instructions

**Comment by Lichtso** [comment]
> @Lichtso has suggested an ABI change

This would invert the order of nested instructions but keep the order of sibling instructions and
be in line with what `InstructionRecorder` and `TransactionStatusSender` report. And should make debugging less confusing as well.

> The status field seems unhelpful

The status field can be inferred anyway and is thus redundant, I think: Every edge in `depth` (decrement step) where `depth < get_invoke_depth()` is "InProgress", everything else is "Success".

**Comment by jackcmay** [comment]
> > (depth, status, instruction)
> 
> The status field seems unhelpful. The only way that B could be `InProgress` at that time is due to the `Success`-ful result of the previous instructions

Yup, I updated the earlier proposal

**Comment by mvines** [comment]
sgtm!  Just to be clear though, the last `(0, B)` item would not be present since B is not an inner instruction:
> Then B's processed inner instruction list is: [(1, F), (1, E), (2, D), (1, C), (0, B)]

I do still think for symmetry it would be nice if the syscall also returned outer instructions, as we're just pushing the special casing from a one-time implementation in the runtime to N programs that each may have their own bugs

**Comment by jackcmay** [comment]
> sgtm! Just to be clear though, the last `(0, B)` item would not be present since B is not an inner instruction:
> 
> > Then B's processed inner instruction list is: [(1, F), (1, E), (2, D), (1, C), (0, B)]
> 
> I do still think for symmetry it would be nice if the syscall also returned outer instructions, as we're just pushing the special casing from a one-time implementation in the runtime to N programs that each may have their own bugs

Yeah, kinda weird to include the other top-level instructions and not their inner instructions via this api.

**Comment by mvines** [comment]
> Yeah, kinda weird to include the other top-level instructions and not their inner instructions via this api.

Given:
```
/// A
/// B -> C -> D
/// B -> E
/// B -> F
/// ** B calls `get_processed_inner_instruction()` here **
/// G

```

Dropping D entirely, and only returning `[(1, F), (1, E), (1, C)]` would be fine for the SPL Token use case.  We don't actually care about D, as it's an implementation detail of C.

It would be lovely if G called `get_processed_inner_instructions()` and result was `[(0,B),(0,A)]`.  That is, rename the syscall to `get_processed_instructions()`.

**Comment by mvines** [comment]
> Just report processed instructions at the current depth, then `get_processed_instructions()` would return `[B, A]` and depth could be discovered via `sol_get_invoke_depth`

Yeah. I guess if/when we ran into a need to report ALL the processed instructions, a new `get_all_processed_instructions()` syscall could be added. Or we could reserve space in the `get_processed_instructions()` args and return values to allow this to be implemented in the future without a new syscall. I don't have strong thoughts about either at the moment.

> Would probably be helpful to also return the current instruction without having to refer to the sysvar

The instruction that's currently executing? Isn't that already known because it's literally executing at that moment? I'm probably not understanding :)

**Comment by jackcmay** [comment]
> It would be lovely if G called `get_processed_inner_instructions()` and result was `[(0,B), (0,A)]`. That is, rename the syscall to `get_processed_instructions()`.

And B calling it would return `[(1, F), (1, E), (1, C), (0, A)]`?

**Comment by jackcmay** [comment]
> The instruction that's currently executing? Isn't that already known because it's literally executing at that moment? I'm probably not understanding :)

Yeah, nvmd, that's communicated via the current program's arguments

**Comment by mvines** [comment]
> And B calling it would return `[(1, F), (1, E), (1, C), (0, A)]`?

From this call sight, yes:
```
/// A
/// A -> X
/// A -> Y-> Z
/// B -> C -> D
/// B -> E
/// B -> F
/// ** B calls `get_processed_instruction()` here **
/// G
```

**Comment by mvines** [comment]
I mean this would also be cool to return as well and would be enable more use cases beyond what SPL Token requires I guess:
`[(1, F), (1, E), (2, D), (1, C), (2, Z), (1, Y), (1, X), (0, A)]`?

**Comment by Lichtso** [comment]
This would be the entire trace of all instructions (outer and inner).

I think there is a "B" missing in that example like so: `[(1, F), (1, E), (2, D), (1, C), (0, B), (2, Z), (1, Y), (1, X), (0, A)]` , right?

**Comment by jackcmay** [comment]
> I mean this would also be cool to return as well and would be enable more use cases beyond what SPL Token requires I guess: `[(1, F), (1, E), (2, D), (1, C), (2, Z), (1, Y), (1, X), (0, A)]`

Could pass a max depth:
0 would return `[(0, A)]`
1 would return `[(1, F), (1, E), (1, C), (0, A)]`
2 would return `[(1, F), (1, E), (2, D), (1, C), (2, Z), (1, Y), (1, X), (0, A)]`

**Comment by jackcmay** [comment]
> This would be the entire trace of all instructions (outer and inner).
> 
> I think there is a "B" missing in that example like so: `[(1, F), (1, E), (2, D), (1, C), (0, B), (2, Z), (1, Y), (1, X), (0, A)]` , right?

That goes back to the question of processed or not, if processed no, if all instructions, yes.

**Comment by mvines** [comment]
> I think there is a "B" missing in that example like so: `[(1, F), (1, E), (2, D), (1, C), (0, B), (2, Z), (1, Y), (1, X), (0, A)]` , right?

right, i missed `(0,B)` there :)

**Comment by Lichtso** [comment]
As I said, we should always include all instructions. It is easy to filter out the unfinished ones if a program does not want that.

**Comment by mvines** [comment]
> Could pass a max depth:

👍🏼

**Comment by mvines** [comment]
> As I said, we should always include all instructions. It is easy to filter out the unfinished ones if a program does not want that.

The future inner instructions are not known though, so the runtime could only return the unfinished transaction instructions and we already have the instructions sysvar for can be used for that purpose.

**Comment by jackcmay** [comment]
I'm a little uneasy about `[(1, F), (1, E), (2, D), (1, C), (0, B), (2, Z), (1, Y), (1, X), (0, A)]` where all the instructions are processed but B, and the caller has to distinguish that B is the not-yet-fully-processed instruction in the list.

**Comment by mvines** [comment]
> I'm a little uneasy about `[(1, F), (1, E), (2, D), (1, C), (0, B), (2, Z), (1, Y), (1, X), (0, A)]` where all the instructions are processed but B, and the caller has to distinguish that B is the not-yet-fully-processed instruction in the list.

🤔 yeah

```
/// A
/// A -> X
/// A -> Y -> Z
///                ** Z calls `get_processed_instruction()` **
/// B
```
At this call site, I think the result would currently be `[(1, Y), (1, X), (0, A)]` or `[(2, Z), (1, X), (0, A)]`?  If only fully processed instructions were returned then Z would see a result of `[(1,X)]`.  The former seems neat as it enables an invoked program to figure out who's calling it in all cases

**Comment by jackcmay** [comment]
> At this call site, I think the result would currently be `[(1, Y), (1, X), (0, A)]` or `[(2, Z), (1, X), (0, A)]`? 

Result would be `[(1, Y), (1, X), (0, A)]`

**Comment by mvines** [comment]
Result would be empty if the syscall was called before A invoked X too?

**Comment by jackcmay** [comment]
> Result would be empty if the syscall was called before A invoked X too?

The current implementation yes, the proposed would return `[(0, A)]`

**Comment by jackcmay** [comment]
@mvines @CriesofCarrots , given:
```
/// A
/// A -> X
/// A -> Y -> Z
///                ** Z calls `get_processed_instruction()` **
/// B
```
If SPL token is Z, will returning `[]` be expected, or does SPL token need X?  And if needs X, does it also need A?

**Comment by jackcmay** [comment]
@Lichtso One of the impedments to relying directly on `InstructionRecorder` and something we need to fix before ABIv2 is this: https://github.com/solana-labs/solana/issues/19467

**Comment by mvines** [comment]
> @mvines @CriesofCarrots , given:
> 
> ```
> /// A
> /// A -> X
> /// A -> Y -> Z
> ///                ** Z calls `get_processed_instruction()` **
> /// B
> ```
> 
> If SPL token is Z, will returning `[]` be expected, or does SPL token need X? And if needs X, does it also need A?

The case we'd need is when SPL Token is `Y` and it needs to see `X`.

But for the sake of discussion, if SPL token was Z (or Y) then I'd expect it to see X since X has completed.  I think what we're trying to determine is whether Z would also see Y and A since those instructions haven't completed yet. 

I'm leaning towards including Y and A, even though they are in progress, since this allows Z to figure out who's calling it and I think that enables some interesting use cases (programs changing their behaviour based on caller).  But perhaps that's actually an anti-pattern/footgun that we shouldn't enable?

**Comment by jackcmay** [comment]
The reason I ask is that X isn't a direct sibling of Z within Y.

Seems like a separate syscall should be added, something like `get_invoke_stack` to get the callers, and later `get_all_instructions` to get all the instructions called so far.  Some of these could be utility routines over the updated `InstructionRecorder` that may be exposed in ABIv2

**Comment by mvines** [comment]
Got it.  Yeah not opposed to deferring a more general solution to later.  SPL Token just needs to inspect the previous sibling instruction at its depth.

**Comment by CriesofCarrots** [comment]
Hey, all. Sorry I was offline for most of this dialog. Is the PR up to date with the currently proposed implementation? If so, I'll catch up with a close read.
Based on above, it sounds like it might be possible to have this syscall work for top-level instructions after all? 🙏 I hope so. I still think it will be better to have the api consistent for instructions at any depth in the call stack.

**Comment by jackcmay** [comment]
@CriesofCarrots Yeah, I see the desire for this api to include top-level instructions and that becomes non-trivial to define the desired behavior.  For example, if we include top-level instructions do we also include their inner instructions?   Do we include all instructions a depth minus?  Do we include instructions currently being processed and let the program figure out via depth if they are in-progress or not?  A lot of info was shared in this thread, reading through that would be helpful.  My current thinking is for this API to include all inner instructions of the current top-level and take a max depth.  Then we could provide a separate syscall to get the call stack.  I want to make sure we cover the use case for SPL but also provide a clear API for other use cases without clouding or returning the world.  I think there is also an argument to include all instructions at depth or below that have been processed which would include processed top-level instructions.

**Comment by CriesofCarrots** [comment]
Okay, I've read through everything carefully now.
If we take the general approach, I think that we do need to include instructions currently being processed. Otherwise, it may not be clear which lower level ix called a higher one. eg, in the example way above, the output `[(1, F), (1, E), (2, D), (1, C), (2, Z), (1, Y), (1, X), (0, A)]` could be from either of these call trees (or one of many other combinations):

```
/// A
/// A -> X
/// A -> Y -> Z
/// B -> C -> D
/// B -> E
/// B -> F
/// ** B calls `get_processed_instruction()` here **
/// G

/// A
/// A -> X
/// A -> Y -> Z
/// A -> C -> D
/// A -> E
/// A -> F
/// B
/// ** B calls `get_processed_instruction()` here **
/// G
```

I think that ambiguity could be a big footgun.

> Yeah not opposed to deferring a more general solution to later. SPL Token just needs to inspect the previous sibling instruction at its depth.

That said, this ^ . It does appeal to me to implement the simpler approach now, and offer the more general utility later if needed. That simpler implementation being:

```
/// A
/// A -> X
/// A -> Y -> Z
/// B -> C -> D
/// B -> E
/// B -> F
/// ** F calls `get_processed_instruction()` here **
/// G
/// ** G calls `get_processed_instruction()` here **
```

F sees: `[(1, E), (1, C)]` or `[E, C]`
B sees: `[(0, B), (0, A)]` or `[B, A]`

**Comment by mvines** [comment]
Maybe we need 

> ```
> /// A
> /// A -> X
> /// A -> Y -> Z
> /// B -> C -> D
> /// B -> E
> /// B -> F
> /// ** F calls `get_processed_instruction()` here **
> /// G
> /// ** G calls `get_processed_instruction()` here **
> ```
> 
> 
>     
>   
> 
> F sees: `[(1, E), (1, C)]` or `[E, C]`
> B sees: `[(0, B), (0, A)]` or `[B, A]`

"B sees" -> "G sees" right?  Sounds good to me if so!

But maybe we need a call with everybody on the PR to work out any remaining design discussion with our vocal folds?

**Comment by CriesofCarrots** [comment]
> "B sees" -> "G sees" right? Sounds good to me if so!

Oops, yep, edited in my comment

> But maybe we need a call with everybody on the PR to work out any remaining design discussion with our vocal folds?

Happy to oscillate some vocal folds. I think Jack is unavailable the rest of this week, so maybe Monday?

**Comment by jackcmay** [comment]
In the examples above I'm inclined to return `[]` from call depth 0 since there are no sibling instructions (just top-level instructions).  The main reason for doing that is I think it is confusing to return `F sees: [E,C]` and not include the top-level successfully processed instruction A (`[E, C, A]`).  A has been processed and so it come down to what is the definition of this API, is it "sibling instructions" or "processed instructions", or maybe going with @CriesofCarrots 's suggestion "processed instructions at my call depth"?

**Comment by mvines** [comment]
The previously (successfully -- implicit) processed instructions at the caller's depth is what we're immediately after.
At depth 0, this set would be the previously processed top-level instructions.

**Comment by jackcmay** [comment]
Replaced by #22859

**Comment by jackcmay** [comment]
Replaced by https://github.com/solana-labs/solana/pull/22859

**Pull request closed without merging**
